src/main/java/xyz/nucleoid/plasmid/util/BucketScanner.java

```diff
@@ -0,0 +1,253 @@
+     */
+    public static Builder create(RuleTest ruleTest, ServerWorld world, Random random) {
+        return create((previousPos, nextPos) -> ruleTest.test(world.getBlockState(nextPos), random));
+    }
+
+    private void initialize() {
+        int amount = 1;
+        while(amount != builder.maxAmount) {
+            if(scanQueue.isEmpty()) {
+                return;
+            }
+            long pos;
+            switch(builder.searchType) {
+                case BREADTH_FIRST:
+                default:
+                    pos = scanQueue.pollLast();
+                    break;
+                case DEPTH_FIRST:
+                    pos = scanQueue.pollFirst();
+                    break;
+            }
+            if(builder.connectivity == Connectivity.TWENTY_SIX) {
+                for(int i = -1; i <= 1; i += 2) {
+                    for(int j = -1; j <= 1; j += 2) {
+                        for(int k = -1; k <= 1; k += 2) {
+                            scan(pos, i, j, k);
+                        }
+                    }
+                }
+            }
+            if(builder.connectivity != Connectivity.SIX) {
+                for(int i = -1; i <= 1; i += 2) {
+                    for(int j = -1; j <= 1; j += 2) {
+                        scan(pos, i, j, 0);
+                        scan(pos, 0, i, j);
+                        scan(pos, j, 0, i);
+                    }
+                }
+            }
+            for(Direction direction : Direction.values()) {
+                scan(pos, direction.getOffsetX(), direction.getOffsetY(), direction.getOffsetZ());
+            }
+            amount++;
+        }
+    }
+
+    private void scan(long previousLong, int offsetX, int offsetY, int offsetZ) {
+        this.previousMutable.set(previousLong);
+        this.nextMutable.set(previousLong).move(offsetX, offsetY, offsetZ);
+        if(test(previousMutable, nextMutable)) {
+            long nextLong = nextMutable.asLong();
+            int nextDepth = depthMap.get(previousLong) - 1;
+            if(nextDepth > 0) {
+                scanQueue.push(nextLong);
+            }
+            depthMap.put(nextLong, nextDepth);
+        }
+    }
+
+    private boolean test(BlockPos previousPos, BlockPos nextPos) {
+        return this.builder.predicate.test(previousPos, nextPos) && !depthMap.containsKey(nextPos.asLong()) && !scanQueue.contains(nextPos.asLong());
+    }
+
+    /**
+     * @return the set of positions found by the scanner
+     */
+    public LongSet getPositions() {
+        return new LongArraySet(depthMap.keySet());
+    }
+
+    /**
+     * Gives a set of positions depending on the depth values of those positions.
+     *
+     * @param predicate determines if a position should enter the set depending of its depth value
+     * @return the set of positions found by the scanner that verify the predicate
+     */
+    public LongSet getPositions(Predicate<Integer> predicate) {
```

Comment:
I would use an `IntPredicate` or even a custom functional interface here:

```suggestion
    public LongSet getPositions(IntPredicate predicate) {
```

Reply:
I think another custom functional interface might be too much, just for an integer predicate.